reader resolution algorithm seems inconsistent between stdlib and importlib_resources

Hello friends,

I'm looking at how to reimplement resource overrides in Pyramid. It's a unique feature that in the past involved patching a custom `package.__loader__` which could intercept and do its own file lookups for various requests. This worked with all pkg_resources apis, like `pkg_resources.resource_filename`, etc to return a different path than the one physically on disk when a path is overridden.

To implement this, I've been going through the PEPs and trying to understand the protocol that `importlib.resources` uses to support overriding how resources are loaded from a package. I've narrowed this down to `spec.loader.get_resource_reader()` being able to be defined. My plan will be to define a `ProxyLoader` which defines its own `get_resource_reader` but proxies the rest of the loader functionality back to the original loader. So, I'm looking at how to define a custom `abc.TraversableResources` class that does what I need, and then hook `package.__spec__.loader` to return it from `get_resource_reader()`.

**From what I can tell, the logic here is inconsistent between importlib_resources and the stdlib impl in CPython. I'm looking here**:

https://github.com/python/cpython/blob/5ddb2740404ad62e74ecb0ea97ed98651821ed3f/Lib/importlib/resources/_adapters.py#L28-L29

compared to

https://github.com/python/importlib_resources/blob/7d8b0200204e40bd32fa16bffb593c8795462702/importlib_resources/_compat.py#L75-L90

In both cases, TraversableResourceLoader is attempting to determine which reader to return. In the stdlib, preference is given to the native `spec.loader.get_resource_reader()` if the reader returned has a `files()` method. However, importlib_resources inverts the lookup such that zip/namespace/file take priority over the "native" reader. So for example, if `self.path` exists, then whatever I set as `spec.loader.get_resource_reader()` will never be invoked and hence I can't override the lookups.

Currently, to get this to work with importlib_resources, it appears that I would need the `spec.path` to be a non-existent path, which would trigger the logic to fallback to the native reader that I patch, but this is by far not ideal.

The TLDR is that I would propose importlib_resources moves the `_native_reader(self.spec)` invocation to the TOP of the list, which would be more in line with the stdlib.

I would really appreciate any guidance here. Maybe I'm completely off base and missing something critical in the importlib machinery, and happy to talk more about this use-case in Pyramid that has been supported for many years via a PEP320 loader used by pkg_resources. Thank you!

First, thanks for sharing your detailed analysis and considerate question.

I agree, it should be possible to accomplish what you seek with a simple and straightforward hook.

In 3e24498, I explored simply giving the native reader precedence, but it triggers a few test failures on Python 3.10 and later, all in ResourceFromNamespaceZipTests:

```
FAILED importlib_resources/tests/test_resource.py::ResourceFromNamespaceZipTests::test_is_submodule_resource - NotADirectoryError: MultiplexedPath only supports directories
FAILED importlib_resources/tests/test_resource.py::ResourceFromNamespaceZipTests::test_read_submodule_resource_by_name - NotADirectoryError: MultiplexedPath only supports directories
FAILED importlib_resources/tests/test_resource.py::ResourceFromNamespaceZipTests::test_submodule_contents - NotADirectoryError: MultiplexedPath only supports directories
FAILED importlib_resources/tests/test_resource.py::ResourceFromNamespaceZipTests::test_submodule_contents_by_name - NotADirectoryError: MultiplexedPath only supports directories
```

I'll need to dig deeper to understand the meaning behind those failures and what else it might imply about the current approach.

--- a/importlib_resources/future/adapters.py
+++ b/importlib_resources/future/adapters.py
@@ -1,9 +1,42 @@
-import contextlib
+import functools
 import pathlib
 from contextlib import suppress
 from types import SimpleNamespace
 
 from .. import readers, _adapters
+
+
+def _block_standard(reader_getter):
+    """
+    Wrap _adapters.TraversableResourcesLoader.get_resource_reader
+    and intercept any standard library readers.
+    """
+
+    @functools.wraps(reader_getter)
+    def wrapper(*args, **kwargs):
+        """
+        If the reader is from the standard library, return None to allow
+        allow likely newer implementations in this library to take precedence.
+        """
+        try:
+            reader = reader_getter(*args, **kwargs)
+        except NotADirectoryError:
+            # MultiplexedPath may fail on zip subdirectory
+            return
+        # Python 3.10+
+        if reader.__class__.__module__.startswith('importlib.resources.'):
+            return
+        # Python 3.8, 3.9
+        if isinstance(reader, _adapters.CompatibilityFiles) and (
+            reader.spec.loader.__class__.__module__.startswith('zipimport')
+            or reader.spec.loader.__class__.__module__.startswith(
+                '_frozen_importlib_external'
+            )
+        ):
+            return
+        return reader
+
+    return wrapper
 
 
 class TraversableResourcesLoader(_adapters.TraversableResourcesLoader):
@@ -16,18 +49,10 @@ class TraversableResourcesLoader(_adapters.TraversableResourcesLoader):
     """
 
     def get_resource_reader(self, name):
-        with contextlib.suppress(Exception):
-            return self._block_standard(super().get_resource_reader(name))
-        return self._standard_reader()
-
-    def _block_standard(self, reader):
-        """
-        If the reader is from the standard library, raise an exception to
-        allow likely newer implementations in this library to take precedence.
-        """
-        if reader.__class__.__module__.startswith('importlib.resources.'):
-            raise RuntimeError("Reader blocked to be superseded.")
-        return reader
+        return (
+            _block_standard(super().get_resource_reader)(name)
+            or self._standard_reader()
+        )
 
     def _standard_reader(self):
         return self._zip_reader() or self._namespace_reader() or self._file_reader()

--- a/importlib_resources/future/adapters.py
+++ b/importlib_resources/future/adapters.py
@@ -1,3 +1,4 @@
+import contextlib
 import pathlib
 from contextlib import suppress
 from types import SimpleNamespace
@@ -15,7 +16,18 @@ class TraversableResourcesLoader(_adapters.TraversableResourcesLoader):
     """
 
     def get_resource_reader(self, name):
-        return self._standard_reader() or super().get_resource_reader(name)
+        with contextlib.suppress(Exception):
+            return self._block_standard(super().get_resource_reader(name))
+        return self._standard_reader()
+
+    def _block_standard(self, reader):
+        """
+        If the reader is from the standard library, raise an exception to
+        allow likely newer implementations in this library to take precedence.
+        """
+        if reader.__class__.__module__.startswith('importlib.resources.'):
+            raise RuntimeError("Reader blocked to be superseded.")
+        return reader
 
     def _standard_reader(self):
         return self._zip_reader() or self._namespace_reader() or self._file_reader()

--- a/newsfragments/295.feature.rst
+++ b/newsfragments/295.feature.rst
@@ -1,0 +1,1 @@
+Future compatibility adapters now ensure that standard library readers are replaced without overriding non-standard readers.

